File: Java/39.Combination_Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        backtrack(candidates, target, 0, new ArrayList<>(), ans);
        return ans;
    }

    private void backtrack(int[] candidates, int target, int index, List<Integer> prefix, List<List<Integer>> ans) {
        if (target < 0) {
            return;
        } else if (target == 0) {
            ans.add(new ArrayList<>(prefix));
        } else {
            for (int i = index; i < candidates.length; ++i) {
                prefix.add(candidates[i]);
                backtrack(candidates, target - candidates[i], i, prefix, ans);
                prefix.remove(prefix.size() - 1);
            }
        }
    }
}

File: Python3/39.Combination_Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        sorted_candidates = sorted(candidates) #not required 
        self.backtrack(sorted_candidates, target, 0, list(), ans)
        return ans
    
    def backtrack(self, candidates, target, index, prefix, ans):
        if target < 0:
            return
        if target == 0:
            ans.append(prefix.copy())
            return
        for i in range(index, len(candidates)):
            prefix.append(candidates[i])
            self.backtrack(candidates, target - candidates[i], i, prefix, ans)
            prefix.pop()